proposals/async-method-builders.md

```diff
@@ -0,0 +1,172 @@
+
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+* The syntax for applying such an attribute to a method is verbose.  This is an advanced feature, but a developer using it frequently could create their own attribute derived from the one they care about, and then use that derived attribute to simplify, e.g.
+```C#
+class Pool<T> : AsyncMethodBuilderAttribute<T>
+{
+    public Pool() : base(typeof(AsyncValueTaskMethodBuilder<T>)) { }
+}
+...
+[Pool]
+internal async ValueTask<int> ExampleAsync() { ... }
+```
+I don't know exactly how this could be made to work with the createArguments support, though.  The derived attribute could also accept a params array and pass it down to the base, but the compiler would need to be able to recognize it.  Maybe a heuristic of special-casing any params object[] at the end of an AsyncMethodBuilder-derived attribute.
+
+## Alternatives
+[alternatives]: #alternatives
+
+- Implement a different task-like type and expose that difference to consumers.  `ValueTask` was made extensible via the `IValueTaskSource` interface to avoid that need, however.
+- Address just the ValueTask pooling part of the issue by enabling the experiment as the on-by-default-and-only implementation.  That doesn't address other aspects, such as configuring the pooling, or enabling someone else to provide their own builder.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+1. **Attribute.** Should we reuse `[AsyncMethodBuilder(typeof(...))]` or introduce yet another attribute?
+2. **Replace or also create.** All of the examples in this proposal are about replacing a buildable task-like's builder.  Should the feature be scoped to just that? Or should you be able to use this attribute on a method with a return type that doesn't already have a builder (e.g. some common interface)?  That could impact overload resolution.
+3. **Virtuals / Interfaces.** What is the behavior if the attribute is specified on an interface method?  I think it should either be a nop or a compiler warning/error, but it shouldn't impact implementations of the interface.  A similar question exists for base methods that are overridden, and there again I don't think the attribute on the base method should impact how an override implementation behaves. Note the current attribute has Inherited = false on its AttributeUsage.
+4. **Precedence.** If we wanted to do the module/type-level annotation, we would need to decide on which attribution wins in the case where multiple ones applied (e.g. one on the method, one on the containing module).  We would also need to determine if this would necessitate using a different attribute (see (1) above), e.g. what would the behavior be if a task-like type was in the same scope?  Or if a buildable task-like itself had async methods on it, would they be influenced by the attributed applied to the task-like type to specify its default builder?
```

Comment:
Can we add:

- Should the compiler support non-public async method builders? This is not spec'd today but experimentally we only support `public` ones. Once this is configurable it seems likely we will want to support library authors who have non-public ones they want to use.

Reply:
Done. Thanks.